File: libs/datapipe-app/frontend/src/features/ops/metrics/metricsSchema.ts
import type { MetricColumnGroup, MetricDefinition, MetricsModelRow, MetricsTableSchema } from "../../../types/ops";

const COUNT_KEYS = new Set([
    "images_support",
    "support",
    "TP",
    "FP",
    "FN",
    "images",
    "objects",
    "detections",
]);

function def(
    key: string,
    label: string,
    short: string,
    group: string,
    opts: Partial<MetricDefinition> = {},
): MetricDefinition {
    return {
        key,
        label,
        short_label: short,
        group,
        task_types: ["*"],
        format: "float",
        higher_is_better: true,
        visible_by_default: true,
        ...opts,
    };
}

function classificationGroups(): MetricColumnGroup[] {
    return [
        {
            key: "core",
            label: "Core",
            priority: 1,
            metrics: [
                def("weighted_f1_score", "Primary F1", "F1", "core", { primary: true }),
                def("accuracy", "Accuracy", "Acc", "core"),
            ],
        },
        {
            key: "weighted",
            label: "Weighted",
            priority: 2,
            metrics: [
                def("weighted_f1_score", "Weighted F1", "F1", "weighted"),
                def("weighted_precision", "Weighted P", "P", "weighted"),
                def("weighted_recall", "Weighted R", "R", "weighted"),
            ],
        },
        {
            key: "macro",
            label: "Macro",
            priority: 3,
            metrics: [
                def("macro_f1_score", "Macro F1", "F1", "macro"),
                def("macro_precision", "Macro P", "P", "macro"),
                def("macro_recall", "Macro R", "R", "macro"),
            ],
        },
        {
            key: "no_pseudo",
            label: "No pseudo",
            priority: 4,
            metrics: [
                def("weighted_without_pseudo_classes_f1_score", "W F1", "W-F1", "no_pseudo", { visible_by_default: false }),
                def("weighted_without_pseudo_classes_precision", "W Precision", "W-P", "no_pseudo", { visible_by_default: false }),
                def("weighted_without_pseudo_classes_recall", "W Recall", "W-R", "no_pseudo", { visible_by_default: false }),
                def("macro_without_pseudo_classes_f1_score", "M F1", "M-F1", "no_pseudo", { visible_by_default: false }),
                def("macro_without_pseudo_classes_precision", "M Precision", "M-P", "no_pseudo", { visible_by_default: false }),
                def("macro_without_pseudo_classes_recall", "M Recall", "M-R", "no_pseudo", { visible_by_default: false }),
            ],
        },
    ];
}

function countsGroup(): MetricColumnGroup {
    return {
        key: "counts",
        label: "Counts",
        priority: 90,
        metrics: [
            def("support", "Support", "Sup", "counts", { format: "integer" }),
            def("images_support", "Images", "Img", "counts", { format: "integer" }),
            def("images", "Images", "Img", "counts", { format: "integer" }),
            def("objects", "Objects", "Obj", "counts", { format: "integer" }),
            def("detections", "Detections", "Det", "counts", { format: "integer" }),
            def("TP", "TP", "TP", "counts", { format: "integer" }),
            def("FP", "FP", "FP", "counts", { format: "integer", higher_is_better: false }),
            def("FN", "FN", "FN", "counts", { format: "integer", higher_is_better: false }),
        ],
    };
}

function detectionGroups(): MetricColumnGroup[] {
    return [
        {
            key: "core",
            label: "Core",
            priority: 1,
            metrics: [
                def("mAP50_95", "mAP50-95", "mAP", "core", { primary: true }),
                def("mAP50", "mAP50", "mAP50", "core"),
            ],
        },
        {
            key: "weighted",
            label: "Weighted",
            priority: 2,
            metrics: [
                def("weighted_f1_score", "Weighted F1", "F1", "weighted"),
                def("weighted_precision", "Weighted P", "P", "weighted"),
                def("weighted_recall", "Weighted R", "R", "weighted"),
            ],
        },
        {
            key: "macro",
            label: "Macro",
            priority: 3,
            metrics: [
                def("macro_f1_score", "Macro F1", "F1", "macro"),
                def("macro_precision", "Macro P", "P", "macro"),
                def("macro_recall", "Macro R", "R", "macro"),
            ],
        },
        {
            key: "quality",
            label: "Detection",
            priority: 4,
            metrics: [
                def("precision", "Precision", "P", "quality"),
                def("recall", "Recall", "R", "quality"),
                def("f1_score", "F1", "F1", "quality"),
            ],
        },
    ];
}

export function inferTaskTypeFromMetrics(keys: string[]): string | undefined {
    const set = new Set(keys);
    if (set.has("mAP50") || set.has("mAP50_95")) return "detection";
    if (Array.from(set).some((k) => k.startsWith("pose_"))) return "keypoints";
    if (
        set.has("weighted_f1_score")
        || set.has("macro_f1_score")
        || set.has("weighted_precision")
        || set.has("accuracy")
    ) {
        return "classification";
    }
    return undefined;
}

function isBlobSchema(schema: MetricsTableSchema): boolean {
    const quality = schema.groups.find((g) => g.key === "quality");
    if (quality && quality.metrics.length > 4) return true;
    if (schema.groups.length === 1 && schema.groups[0].metrics.length > 4) return true;
    return false;
}

function filterSchemaToAvailable(
    schema: MetricsTableSchema,
    available: string[],
    opts?: { includeCounts?: boolean },
): MetricsTableSchema {
    const present = new Set(available.filter((k) => opts?.includeCounts || !COUNT_KEYS.has(k)));
    const seen = new Set<string>();
    const groups: MetricColumnGroup[] = [];

    for (const group of schema.groups) {
        const metrics = group.metrics.filter((m) => {
            if (seen.has(m.key)) return false;
            if (!present.has(m.key)) return false;
            seen.add(m.key);
            return true;
        });
        if (metrics.length) {
            groups.push({ ...group, metrics });
        }
    }

    const otherKeys = Array.from(present).filter((k) => !seen.has(k)).sort();
    if (otherKeys.length) {
        groups.push({
            key: "other",
            label: "Other",
            priority: 99,
            metrics: otherKeys.map((key) => def(key, key, key.slice(0, 4), "other", { visible_by_default: false })),
        });
    }

    const primary = schema.primary_metric && present.has(schema.primary_metric)
        ? schema.primary_metric
        : groups[0]?.metrics.find((m) => m.primary)?.key
            ?? groups[0]?.metrics[0]?.key
            ?? "f1_score";

    return { task_type: schema.task_type, primary_metric: primary, groups };
}

function fullCanonicalSchema(taskType: string): MetricsTableSchema {
    const tt = taskType === "cls" ? "classification" : taskType;
    let groups: MetricColumnGroup[];
    let primary = "f1_score";
    if (tt === "classification") {
        groups = [...classificationGroups(), countsGroup()];
        primary = "weighted_f1_score";
    } else if (tt === "detection") {
        groups = [...detectionGroups(), countsGroup()];
        primary = "mAP50_95";
    } else {
        groups = [countsGroup()];
    }
    return { task_type: tt, primary_metric: primary, groups };
}

function canonicalSchema(taskType: string, available: string[]): MetricsTableSchema {
    const tt = taskType === "cls" ? "classification" : taskType;
    let groups: MetricColumnGroup[];
    let primary = "f1_score";
    if (tt === "classification") {
        groups = classificationGroups();
        primary = "weighted_f1_score";
    } else if (tt === "detection") {
        groups = detectionGroups();
        primary = "mAP50_95";
    } else {
        groups = [];
    }
    return filterSchemaToAvailable({ task_type: tt, primary_metric: primary, groups }, available);
}

export function buildMetricSchema(
    taskType: string | undefined,
    availableMetrics: string[],
    backendSchema?: MetricsTableSchema,
): MetricsTableSchema {
    const available = availableMetrics.filter((k) => !COUNT_KEYS.has(k));
    const inferred =
        (taskType && taskType !== "auto" ? taskType : undefined)
        ?? inferTaskTypeFromMetrics(available)
        ?? (backendSchema?.task_type !== "custom" ? backendSchema?.task_type : undefined)
        ?? "classification";

    if (inferred === "classification" || inferred === "detection" || inferred === "cls") {
        return canonicalSchema(inferred, available);
    }

    if (backendSchema && !isBlobSchema(backendSchema)) {
        return filterSchemaToAvailable(backendSchema, available);
    }

    return canonicalSchema("classification", available);
}

export type MetricsViewMode = "detailed" | "all";

const ALL_GROUP_ORDER = ["core", "weighted", "macro", "no_pseudo", "quality", "localization", "pose", "task_specific", "counts", "other"];

function sortGroups(groups: MetricColumnGroup[]): MetricColumnGroup[] {
    return [...groups].sort(
        (a, b) => (ALL_GROUP_ORDER.indexOf(a.key) + 1 || 99) - (ALL_GROUP_ORDER.indexOf(b.key) + 1 || 99),
    );
}

export function collectRowMetricKeys(rows: MetricsModelRow[]): string[] {
    const keys = new Set<string>();
    for (const row of rows) {
        for (const [k, v] of Object.entries(row.metrics ?? {})) {
            if (v != null && typeof v === "number") keys.add(k);
        }
    }
    return Array.from(keys);
}

export function buildAllMetricsSchema(
    baseSchema: MetricsTableSchema,
    rows: MetricsModelRow[],
): MetricsTableSchema {
    const keys = collectRowMetricKeys(rows);
    if (!keys.length) return baseSchema;
    const full = fullCanonicalSchema(baseSchema.task_type);
    return filterSchemaToAvailable(full, keys, { includeCounts: true });
}

const DETAILED_GROUP_ORDER = ["core", "weighted", "macro", "no_pseudo", "quality", "localization", "pose", "task_specific"];

export function visibleGroups(
    schema: MetricsTableSchema,
    viewMode: MetricsViewMode,
): MetricColumnGroup[] {
    if (viewMode === "all") {
        return sortGroups(schema.groups);
    }
    const sorted = [...schema.groups].sort(
        (a, b) => (DETAILED_GROUP_ORDER.indexOf(a.key) + 1 || 99) - (DETAILED_GROUP_ORDER.indexOf(b.key) + 1 || 99),
    );
    return sorted.filter(
        (g) => !["other", "counts", "quality"].includes(g.key) || g.metrics.length <= 3,
    ).filter((g) => g.key !== "other");
}

export function primaryMetricDef(schema: MetricsTableSchema): MetricDefinition | undefined {
    for (const group of schema.groups) {
        const found = group.metrics.find((m) => m.primary || m.key === schema.primary_metric);
        if (found) return found;
    }
    return schema.groups[0]?.metrics[0];
}

export function groupMetricItems(
    rowMetrics: Record<string, number | null | undefined> | undefined,
    group: MetricColumnGroup,
    limit = 3,
) {
    return group.metrics
        .filter((m) => rowMetrics?.[m.key] != null)
        .slice(0, limit)
        .map((m) => ({
            label: m.short_label,
            value: rowMetrics?.[m.key],
            format: m.format,
        }));
}

File: libs/datapipe-ml/datapipe_ml/observability/metrics_schema.py
from __future__ import annotations

from datapipe_ml.observability.schemas import MetricColumnGroup, MetricDefinition, MetricsTableSchema

_COUNT_KEYS = frozenset(
    {
        "images_support",
        "support",
        "TP",
        "FP",
        "FN",
        "TP_extra_bbox",
        "FP_extra_bbox",
        "FN_extra_bbox",
        "images",
        "objects",
        "detections",
        "false_positives",
        "false_negatives",
    }
)


def _def(
    key: str,
    label: str,
    short: str,
    group: str,
    *,
    task_types: list[str] | None = None,
    fmt: str = "float",
    visible: bool = True,
    primary: bool = False,
    higher: bool = True,
) -> MetricDefinition:
    return MetricDefinition(
        key=key,
        label=label,
        short_label=short,
        group=group,
        task_types=task_types or ["*"],
        format=fmt,
        higher_is_better=higher,
        visible_by_default=visible,
        primary=primary,
    )


def _classification_groups() -> list[MetricColumnGroup]:
    return [
        MetricColumnGroup(
            key="core",
            label="Core",
            priority=1,
            metrics=[
                _def("weighted_f1_score", "Primary F1", "Primary", "core", primary=True),
                _def("accuracy", "Accuracy", "Acc", "core"),
                _def("support", "Support", "Sup", "core", fmt="integer"),
            ],
        ),
        MetricColumnGroup(
            key="weighted",
            label="Weighted",
            priority=2,
            metrics=[
                _def("weighted_f1_score", "Weighted F1", "F1", "weighted"),
                _def("weighted_precision", "Weighted Precision", "P", "weighted"),
                _def("weighted_recall", "Weighted Recall", "R", "weighted"),
            ],
        ),
        MetricColumnGroup(
            key="macro",
            label="Macro",
            priority=3,
            metrics=[
                _def("macro_f1_score", "Macro F1", "F1", "macro"),
                _def("macro_precision", "Macro Precision", "P", "macro"),
                _def("macro_recall", "Macro Recall", "R", "macro"),
            ],
        ),
        MetricColumnGroup(
            key="no_pseudo",
            label="No pseudo",
            priority=4,
            metrics=[
                _def("weighted_without_pseudo_classes_f1_score", "Weighted F1", "F1", "no_pseudo", visible=False),
                _def("weighted_without_pseudo_classes_precision", "Weighted P", "P", "no_pseudo", visible=False),
                _def("weighted_without_pseudo_classes_recall", "Weighted R", "R", "no_pseudo", visible=False),
                _def("macro_without_pseudo_classes_f1_score", "Macro F1", "F1", "no_pseudo", visible=False),
                _def("macro_without_pseudo_classes_precision", "Macro P", "P", "no_pseudo", visible=False),
                _def("macro_without_pseudo_classes_recall", "Macro R", "R", "no_pseudo", visible=False),
            ],
        ),
    ]


def _detection_groups() -> list[MetricColumnGroup]:
    return [
        MetricColumnGroup(
            key="core",
            label="Core",
            priority=1,
            metrics=[
                _def("mAP50_95", "mAP50-95", "mAP", "core", task_types=["detection"], primary=True),
                _def("mAP50", "mAP50", "mAP50", "core", task_types=["detection"]),
                _def("support", "Support", "Sup", "core", fmt="integer"),
            ],
        ),
        MetricColumnGroup(
            key="weighted",
            label="Weighted",
            priority=2,
            metrics=[
                _def("weighted_f1_score", "Weighted F1", "F1", "weighted"),
                _def("weighted_precision", "Weighted Precision", "P", "weighted"),
                _def("weighted_recall", "Weighted Recall", "R", "weighted"),
            ],
        ),
        MetricColumnGroup(
            key="macro",
            label="Macro",
            priority=3,
            metrics=[
                _def("macro_f1_score", "Macro F1", "F1", "macro"),
                _def("macro_precision", "Macro Precision", "P", "macro"),
                _def("macro_recall", "Macro Recall", "R", "macro"),
            ],
        ),
        MetricColumnGroup(
            key="quality",
            label="Detection",
            priority=4,
            metrics=[
                _def("precision", "Precision", "P", "quality", task_types=["detection"]),
                _def("recall", "Recall", "R", "quality", task_types=["detection"]),
                _def("f1_score", "F1", "F1", "quality", task_types=["detection"]),
            ],
        ),
        MetricColumnGroup(
            key="localization",
            label="Localization",
            priority=5,
            metrics=[
                _def("iou_mean", "Mean IoU", "IoU", "localization", task_types=["detection"], visible=False),
            ],
        ),
    ]


def _segmentation_groups() -> list[MetricColumnGroup]:
    return [
        MetricColumnGroup(
            key="core",
            label="Core",
            priority=1,
            metrics=[
                _def("iou_mean", "mIoU", "mIoU", "core", task_types=["segmentation"], primary=True),
                _def("dice", "Dice", "Dice", "core", task_types=["segmentation"]),
                _def("pixel_accuracy", "Pixel accuracy", "PixAcc", "core", task_types=["segmentation"], visible=False),
                _def("support", "Support", "Sup", "core", fmt="integer"),
            ],
        ),
    ]


def _keypoints_groups() -> list[MetricColumnGroup]:
    return [
        MetricColumnGroup(
            key="core",
            label="Core",
            priority=1,
            metrics=[
                _def("pose_mAP50_95", "Pose mAP50-95", "mAP", "core", task_types=["keypoints"], primary=True),
                _def("pose_mAP50", "Pose mAP50", "mAP50", "core", task_types=["keypoints"]),
                _def("support", "Support", "Sup", "core", fmt="integer"),
            ],
        ),
        MetricColumnGroup(
            key="pose",
            label="Pose",
            priority=2,
            metrics=[
                _def("pose_P", "Pose precision", "P", "pose", task_types=["keypoints"]),
                _def("pose_R", "Pose recall", "R", "pose", task_types=["keypoints"]),
                _def("OKS", "OKS", "OKS", "pose", task_types=["keypoints"], visible=False),
                _def("PCK", "PCK", "PCK", "pose", task_types=["keypoints"], visible=False),
            ],
        ),
    ]


def infer_task_type_from_metrics(available: set[str]) -> str | None:
    if available & {"mAP50", "mAP50_95"}:
        return "detection"
    if any(k.startswith("pose_") for k in available):
        return "keypoints"
    if available & {
        "weighted_f1_score",
        "macro_f1_score",
        "weighted_precision",
        "macro_precision",
        "accuracy",
    }:
        return "classification"
    return None


def _heuristic_group(key: str) -> str:
    lk = key.lower()
    if any(x in lk for x in ("support", "count", "items", "images", "objects")):
        return "counts"
    if any(x in lk for x in ("map", "iou", "dice", "oks", "pck", "hausdorff")):
        return "task_specific"
    if any(x in lk for x in ("f1", "precision", "recall", "accuracy")):
        return "quality"
    return "other"


def _fallback_groups(available: set[str], primary_metric: str) -> list[MetricColumnGroup]:
    by_group: dict[str, list[MetricDefinition]] = {}
    for key in sorted(available):
        if key in _COUNT_KEYS:
            continue
        group = _heuristic_group(key)
        by_group.setdefault(group, []).append(
            _def(key, key.replace("_", " ").title(), key[:6], group, primary=(key == primary_metric))
        )
    return [
        MetricColumnGroup(key=g, label=g.replace("_", " ").title(), priority=i + 1, metrics=metrics)
        for i, (g, metrics) in enumerate(sorted(by_group.items(), key=lambda x: x[0]))
    ]


def build_metric_schema(
    task_type: str | None,
    available_metrics: list[str] | set[str],
    *,
    primary_metric: str | None = None,
) -> MetricsTableSchema:
    tt = (task_type or "custom").lower()
    available = set(available_metrics)
    if tt == "custom":
        inferred = infer_task_type_from_metrics(available)
        if inferred:
            tt = inferred
    if tt in ("classification", "cls"):
        groups = _classification_groups()
        pm = primary_metric or "weighted_f1_score"
    elif tt == "detection":
        groups = _detection_groups()
        pm = primary_metric or "mAP50_95"
    elif tt == "segmentation":
        groups = _segmentation_groups()
        pm = primary_metric or "iou_mean"
    elif tt in ("keypoints", "pose"):
        groups = _keypoints_groups()
        pm = primary_metric or "pose_mAP50_95"
    else:
        pm = primary_metric or (next(iter(available), None) or "f1_score")
        groups = _fallback_groups(available, pm)

    filtered: list[MetricColumnGroup] = []
    seen_keys: set[str] = set()
    for group in groups:
        metrics = []
        for m in group.metrics:
            if m.key in seen_keys:
                continue
            if available and m.key not in available:
                continue
            seen_keys.add(m.key)
            metrics.append(m)
        if metrics:
            filtered.append(group.model_copy(update={"metrics": metrics}))

    other_keys = sorted(k for k in available if k not in seen_keys and k not in _COUNT_KEYS)
    if other_keys:
        filtered.append(
            MetricColumnGroup(
                key="other",
                label="Other",
                priority=99,
                metrics=[_def(k, k, k[:6], "other", visible=False) for k in other_keys],
            )
        )

    return MetricsTableSchema(task_type=tt, primary_metric=pm, groups=filtered)
